d2/src/main/java/com/linkedin/d2/balancer/util/canary/BasicCanaryDistributionProviderImpl.java

```diff
@@ -0,0 +1,127 @@
+ * 3) hashing result falls into the ramp scope.
+ * , where hashing result is the absolute value of the hash code of "serviceName+hostName".
+ */
+public class BasicCanaryDistributionProviderImpl implements CanaryDistributionProvider
+{
+  private static final Logger _log = LoggerFactory.getLogger(BasicCanaryDistributionProviderImpl.class);
+
+  private final String _serviceName; // name of the running service/application
+  private final String _hostName; // name of the running host
+
+  public BasicCanaryDistributionProviderImpl(String serviceName, String hostName)
+  {
+    _serviceName = serviceName;
+    _hostName = hostName;
+  }
+
+  @Override
+  public Distribution distribute(D2CanaryDistributionStrategy strategy)
+  {
+    switch (strategy.getStrategy()) {
+      case TARGET_HOSTS:
+        return distributeByTargetHosts(strategy);
+      case TARGET_APPLICATIONS:
+        return distributeByTargetApplications(strategy);
+      case PERCENTAGE:
+        return distributeByPercentage(strategy);
+      case DISABLED:
+        return Distribution.STABLE;
+      default:
+        _log.warn("Invalid distribution strategy type: " + strategy.getStrategy().name());
+        return Distribution.STABLE;
+    }
+  }
+
+  protected Distribution distributeByTargetHosts(D2CanaryDistributionStrategy strategy)
+  {
+    TargetHostsStrategyProperties targetHostsProperties = strategy.getTargetHostsStrategyProperties();
+    if (targetHostsProperties == null) {
+      _log.warn("Empty target hosts properties in distribution strategy type.");
+      return Distribution.STABLE;
+    }
+    return targetHostsProperties.getTargetHosts().stream().anyMatch(this::isHostMatch) ? Distribution.CANARY
+        : Distribution.STABLE;
+  }
+
+  protected Distribution distributeByTargetApplications(D2CanaryDistributionStrategy strategy)
+  {
+    TargetApplicationsStrategyProperties targetAppsProperties = strategy.getTargetApplicationsStrategyProperties();
+    if (targetAppsProperties == null) {
+      _log.warn("Empty target applications properties in distribution strategy type.");
+      return Distribution.STABLE;
+    }
+    return targetAppsProperties.getTargetApplications().stream().anyMatch(this::isServiceMatch) && isCanaryByRampScope(
+        targetAppsProperties.getScope()) ? Distribution.CANARY : Distribution.STABLE;
+  }
+
+  protected Distribution distributeByPercentage(D2CanaryDistributionStrategy strategy)
+  {
+    PercentageStrategyProperties percentageProperties = strategy.getPercentageStrategyProperties();
+    if (percentageProperties == null) {
+      _log.warn("Empty percentage properties in distribution strategy type.");
+      return Distribution.STABLE;
+    }
+    return isCanaryByRampScope(percentageProperties.getScope()) ? Distribution.CANARY : Distribution.STABLE;
+  }
+
+  protected String getServiceName()
+  {
+    return _serviceName == null ? "" : _serviceName;
```

Comment:
Why not do this null prevention logic in the constructor?

Reply:
just wanted to be tolerant with this basic impl, since the hash logic will technically still work will null values (will use empty strings). The container extension uses serviceInfo, so it won't have null values.